tokio/src/net/windows/named_pipe.rs

```diff
@@ -0,0 +1,1166 @@
+///     }
+///
+///     Ok::<_, io::Error>(())
+/// });
+///
+/// /* do something else not server related here */
+/// # Ok(()) }
+/// ```
+///
+/// [`ERROR_PIPE_BUSY`]: crate::winapi::shared::winerror::ERROR_PIPE_BUSY
+/// [Windows named pipe]: https://docs.microsoft.com/en-us/windows/win32/ipc/named-pipes
+#[derive(Debug)]
+pub struct NamedPipe {
+    io: PollEvented<mio_windows::NamedPipe>,
+}
+
+impl NamedPipe {
+    /// Fallibly construct a new named pipe from the specified raw handle.
+    ///
+    /// This function will consume ownership of the handle given, passing
+    /// responsibility for closing the handle to the returned object.
+    ///
+    /// This function is also unsafe as the primitives currently returned have
+    /// the contract that they are the sole owner of the file descriptor they
+    /// are wrapping. Usage of this function could accidentally allow violating
+    /// this contract which can cause memory unsafety in code that relies on it
+    /// being true.
+    ///
+    /// # Errors
+    ///
+    /// This errors if called outside of a [Tokio Runtime] which doesn't have
+    /// [I/O enabled].
+    ///
+    /// [Tokio Runtime]: crate::runtime::Runtime
+    /// [I/O enabled]: crate::runtime::Builder::enable_io
+    pub unsafe fn try_from_raw_handle(handle: RawHandle) -> io::Result<Self> {
+        let named_pipe = mio_windows::NamedPipe::from_raw_handle(handle);
+
+        Ok(NamedPipe {
+            io: PollEvented::new(named_pipe)?,
+        })
+    }
+
+    /// Enables a named pipe server process to wait for a client process to
+    /// connect to an instance of a named pipe. A client process connects by
+    /// creating a named pipe with the same name.
+    ///
+    /// This corresponds to the [`ConnectNamedPipe`] system call.
+    ///
+    /// [`ConnectNamedPipe`]: https://docs.microsoft.com/en-us/windows/win32/api/namedpipeapi/nf-namedpipeapi-connectnamedpipe
+    ///
+    /// ```no_run
+    /// use tokio::net::windows::named_pipe::ServerOptions;
+    ///
+    /// const PIPE_NAME: &str = r"\\.\pipe\mynamedpipe";
+    ///
+    /// # #[tokio::main] async fn main() -> std::io::Result<()> {
+    /// let pipe = ServerOptions::new().create(PIPE_NAME)?;
+    ///
+    /// // Wait for a client to connect.
+    /// pipe.connect().await?;
+    ///
+    /// // Use the connected client...
+    /// # Ok(()) }
+    /// ```
+    pub async fn connect(&self) -> io::Result<()> {
+        loop {
+            match self.io.connect() {
+                Ok(()) => break,
+                Err(e) if e.kind() == io::ErrorKind::WouldBlock => {
+                    self.io.registration().readiness(Interest::WRITABLE).await?;
+                }
+                Err(e) => return Err(e),
+            }
+        }
+
+        Ok(())
+    }
+
+    /// Disconnects the server end of a named pipe instance from a client
+    /// process.
+    ///
+    /// ```
+    /// use tokio::io::AsyncWriteExt as _;
```

Comment:
We don't do this in our examples.
```suggestion
    /// use tokio::io::AsyncWriteExt;
```